tcms/core/tests/test_ajax.py

```diff
@@ -229,3 +233,153 @@ def _assert_tags(self, response):
+        case_tag_counter = _TagCounter('num_cases', test_case_tags)
+        count_for_tag_one = case_tag_counter.calculate_tag_count(self.tag_one)
+        count_for_tag_two = case_tag_counter.calculate_tag_count(self.tag_two)
+
+        self.assertEqual(count_for_tag_one, 0)
+        self.assertEqual(count_for_tag_two, 0)
+
+    def test_with_tag_not_in_query(self):
+        """ Given a QuerySet that does not contain a given tag, the count for this tag should be 0 """
+
+        test_case = TestCaseFactory()
+        test_case.add_tag(self.tag_one)
+
+        test_case_tags = TestCaseTag.objects.filter(
+            tag=self.tag_one).values('tag').annotate(
+            num_cases=Count('tag')).order_by('tag')
+
+        case_tag_counter = _TagCounter('num_cases', test_case_tags)
+        count_for_tag_one = case_tag_counter.calculate_tag_count(self.tag_one)
+        count_for_tag_two = case_tag_counter.calculate_tag_count(self.tag_two)
+
+        self.assertEqual(count_for_tag_one, 1)
+        self.assertEqual(count_for_tag_two, 0)
+
+    def test_with_multiple_tags_and_multiple_objects(self):
+
+        test_plan = TestPlanFactory()
+        test_case = TestCaseFactory()
+        test_run = TestRunFactory()
+
+        test_plan.add_tag(self.tag_one)
+        test_plan.add_tag(self.tag_two)
+        test_plan.add_tag(self.tag_three)
+
+        test_case.add_tag(self.tag_one)
+        test_case.add_tag(self.tag_two)
+
+        test_run.add_tag(self.tag_one)
+
+        test_plan_tags = TestPlanTag.objects.filter(tag__in=self.tags).values('tag').annotate(
+            num_plans=Count('tag')).order_by('tag')
+        test_case_tags = TestCaseTag.objects.filter(tag__in=self.tags).values('tag').annotate(
+            num_cases=Count('tag')).order_by('tag')
+        test_run_tags = TestRunTag.objects.filter(tag__in=self.tags).values('tag').annotate(
+            num_runs=Count('tag')).order_by('tag')
+
+        plan_counter = _TagCounter('num_plans', test_plan_tags)
+        case_counter = _TagCounter('num_cases', test_case_tags)
+        run_counter = _TagCounter('num_runs', test_run_tags)
+
+        self.assertEqual(plan_counter.calculate_tag_count(self.tag_one), 1)
+        self.assertEqual(plan_counter.calculate_tag_count(self.tag_two), 1)
+        self.assertEqual(plan_counter.calculate_tag_count(self.tag_three), 1)
+
+        self.assertEqual(case_counter.calculate_tag_count(self.tag_one), 1)
+        self.assertEqual(case_counter.calculate_tag_count(self.tag_two), 1)
+        self.assertEqual(case_counter.calculate_tag_count(self.tag_three), 0)
+
+        self.assertEqual(run_counter.calculate_tag_count(self.tag_one), 1)
+        self.assertEqual(run_counter.calculate_tag_count(self.tag_two), 0)
+        self.assertEqual(run_counter.calculate_tag_count(self.tag_two), 0)
+
+    def test_in_loop(self):
+
+        tag_one = TagFactory()
+        tag_two = TagFactory()
+        tag_three = TagFactory()
+        tags = [tag_one, tag_two, tag_three]
```

Comment:
These are already created as class attributes in the setup above, why duplicate the setup ?